File: nn/chargpt.ts
import { Tensor } from './tensor';
import { GPT } from './gpt';
import { Adam } from './nn';

export function train_model(
  model: GPT,
  optimizer: Adam,
  docs: string[],
  uchars: string[],
  BOS: number,
  block_size: number,
  num_steps = 1000
): GPT {
  const params = model.parameters();

  for (let step = 0; step < num_steps; step++) {
    const doc = docs[step % docs.length];
    const tokens = [BOS, ...doc.split('').map(ch => uchars.indexOf(ch)), BOS];
    const n = Math.min(block_size, tokens.length - 1);

    const x = [tokens.slice(0, n)];
    const y = [tokens.slice(1, n + 1)];

    optimizer.zeroGrad();
    const { logits } = model.forward(Tensor.from(x));

    const loss = logits.cross_entropy(Tensor.from(y));
    loss.backward();

    let total_norm = 0;
    for (const p of params) {
      if (p.grad) {
        for (const row of p.grad) {
          for (const g of row) {
            total_norm += g * g;
          }
        }
      }
    }
    total_norm = Math.sqrt(total_norm);

    const max_norm = 1.0;
    if (total_norm > max_norm) {
      const clip_coef = max_norm / (total_norm + 1e-6);
      for (const p of params) {
        if (p.grad) {
          p.grad = p.grad.map((row: number[]) => row.map((g: number) => g * clip_coef));
        }
      }
    }

    optimizer.step();
    (optimizer as unknown as { lr: number }).lr = 0.01 * (1 - step / num_steps);

    console.log(`step ${String(step + 1).padStart(4)} / ${num_steps} | loss ${(loss.data[0]?.[0] ?? 0).toFixed(4)}`);
  }

  return model;
}

export function generate_samples(
  model: GPT,
  uchars: string[],
  BOS: number,
  vocab_size: number,
  block_size: number,
  num_samples = 20,
  temperature = 0.5
): string[] {
  console.log('\n--- inference (new, hallucinated names) ---');
  const results: string[] = [];

  for (let sample_idx = 0; sample_idx < num_samples; sample_idx++) {
    let current_token = BOS;
    const sample: string[] = [];
    let kv_caches: { k: Tensor; v: Tensor }[] | undefined;

    for (let pos_id = 0; pos_id < block_size; pos_id++) {
      const x = [[current_token]];
      const { logits, caches } = model.forward(Tensor.from(x), kv_caches);
      kv_caches = caches;

      const last_logits: number[] = logits.data[0] ?? [];
      const max_log = Math.max(...last_logits);
      const exps = last_logits.map((v: number) => Math.exp((v - max_log) / temperature));
      const sum_exp = exps.reduce((a: number, b: number) => a + b, 0);
      const probs = exps.map((v: number) => v / sum_exp);

      let cumsum = 0;
      const r = Math.random();
      let next_token = 0;
      for (let v = 0; v < vocab_size; v++) {
        cumsum += probs[v] ?? 0;
        if (r <= cumsum) {
          next_token = v;
          break;
        }
      }

      if (next_token === BOS) break;
      sample.push(uchars[next_token] ?? '');
      current_token = next_token;
    }

    const generated_name = sample.join('');
    results.push(generated_name);
    console.log(`sample ${String(sample_idx + 1).padStart(2)}: ${generated_name}`);
  }

  return results;
}

File: nn/chargpt.py
import numpy as np
from .tensor import Tensor
from .nn import Adam
from .gpt import GPT

"""
  訓練 GPT 模型。
  訓練時不使用 KV Cache，直接計算整段序列的 loss。
  加入梯度裁剪（Gradient Clipping）防止梯度爆炸。
"""
def train_model(model, optimizer, docs, uchars, BOS, block_size, num_steps=1000):
    params = model.parameters()
    
    for step in range(num_steps):
        doc = docs[step % len(docs)]
        tokens = [BOS] + [uchars.index(ch) for ch in doc] + [BOS]
        n = min(block_size, len(tokens) - 1)
        
        x = np.array([tokens[:n]], dtype=int)
        y = np.array([tokens[1:n+1]], dtype=int)
        
        optimizer.zero_grad()
        logits, _ = model(x, kv_caches=None)  # 訓練不用 KV Cache
        
        loss = logits.cross_entropy(y)
        loss.backward()
        
        # 梯度裁剪：限制所有參數梯度的 L2 範數不超過 max_norm
        max_norm = 1.0
        total_norm = np.sqrt(sum(np.sum(p.grad ** 2) for p in params))
        if total_norm > max_norm:
            clip_coef = max_norm / (total_norm + 1e-6)
            for p in params:
                p.grad *= clip_coef
                
        optimizer.step()
        optimizer.lr = 0.01 * (1 - step / num_steps)
        
        print(f"step {step+1:4d} / {num_steps:4d} | loss {loss.data:.4f}", end='\r')
    
    print()
    return model

"""
  推論函數：使用 KV Cache 加速自回歸生成。
  每次只輸入上一步產生的 token（長度 1），
  快取會自動累積歷史資訊。
"""
def generate_samples(model, uchars, BOS, vocab_size, block_size, num_samples=20, temperature=0.5):
    print("\n--- inference (new, hallucinated names) ---")
    results = []
    
    for sample_idx in range(num_samples):
        current_token = BOS
        sample = []
        kv_caches = None  # 每個新樣本清空快取
        
        for pos_id in range(block_size):
            x = np.array([[current_token]], dtype=int)
            logits, kv_caches = model(x, kv_caches)
            
            last_logits = logits.data[0, 0, :]  # 取當前 token 的 logits
            
            # Temperature Scaling + Softmax
            exps = np.exp(last_logits / temperature - np.max(last_logits / temperature))
            probs = exps / np.sum(exps)
            
            current_token = np.random.choice(range(vocab_size), p=probs)
            if current_token == BOS:
                break
            sample.append(uchars[current_token])
            
        generated_name = ''.join(sample)
        results.append(generated_name)
        print(f"sample {sample_idx+1:2d}: {generated_name}")
        
    return results
